src/main/java/software/amazon/nio/spi/s3/S3DirectoryStream.java

```diff
@@ -0,0 +1,82 @@
+package software.amazon.nio.spi.s3;
+
+import io.reactivex.rxjava3.core.Flowable;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import software.amazon.awssdk.services.s3.model.CommonPrefix;
+import software.amazon.awssdk.services.s3.model.S3Object;
+import software.amazon.awssdk.services.s3.paginators.ListObjectsV2Publisher;
+
+import java.io.IOException;
+import java.nio.file.DirectoryStream;
+import java.nio.file.FileSystem;
+import java.nio.file.Path;
+import java.util.Iterator;
+
+import static software.amazon.nio.spi.s3.Constants.PATH_SEPARATOR;
+
+class S3DirectoryStream implements DirectoryStream<Path> {
+    private final Logger logger = LoggerFactory.getLogger(this.getClass().getName());
+    private final Iterator<Path> iterator;
+
+    S3DirectoryStream(S3FileSystem fs, String bucketName, String finalDirName, Filter<? super Path> filter) {
+        final var listObjectsV2Publisher = fs.client().listObjectsV2Paginator(req -> req
+                .bucket(bucketName)
+                .prefix(finalDirName)
+                .delimiter(PATH_SEPARATOR));
+
+        iterator = pathIteratorForPublisher(filter, fs, finalDirName, listObjectsV2Publisher);
+        //noinspection ResultOfMethodCallIgnored
+        iterator.hasNext();
+    }
+
+    @Override
+    @NonNull
+    public Iterator<Path> iterator() {
+        return iterator;
+    }
+
+    @Override
+    public void close() {}
+
+    /**
+     * Get an iterator for a {@code ListObjectsV2Publisher}. This method is protected level access only for testing
```

Comment:
Can change the java doc to remove the comment about protected level access for testing.